File: bioinforw-data-plotting/assets/bioinforw_ngplot_dump/catalog_loader.R
library(jsonlite)

ngplot_base_dir <- function(base_dir = NULL) {
  if (is.null(base_dir)) {
    normalizePath(dirname(sys.frame(1)$ofile %||% "catalog_loader.R"), winslash = "/", mustWork = FALSE)
  } else {
    normalizePath(base_dir, winslash = "/", mustWork = FALSE)
  }
}

`%||%` <- function(x, y) {
  if (is.null(x)) y else x
}

load_ngplot_catalog <- function(base_dir = NULL) {
  base <- ngplot_base_dir(base_dir)
  fromJSON(file.path(base, "catalog.json"), simplifyVector = FALSE)
}

list_ngplot_templates <- function(base_dir = NULL) {
  load_ngplot_catalog(base_dir)$templates
}

find_ngplot_templates <- function(keyword = NULL, category = NULL, has_python = NULL, has_r = NULL, base_dir = NULL) {
  items <- list_ngplot_templates(base_dir)
  keep <- rep(TRUE, length(items))

  if (!is.null(keyword) && nzchar(keyword)) {
    needle <- tolower(keyword)
    keep <- keep & vapply(items, function(item) {
      hay <- paste(item$name, item$id, item$category, item$categoryTrailText)
      grepl(needle, tolower(hay), fixed = TRUE)
    }, logical(1))
  }

  if (!is.null(category) && nzchar(category)) {
    keep <- keep & vapply(items, function(item) identical(item$category, category), logical(1))
  }

  if (!is.null(has_python)) {
    keep <- keep & vapply(items, function(item) identical(isTRUE(item$hasPython), has_python), logical(1))
  }

  if (!is.null(has_r)) {
    keep <- keep & vapply(items, function(item) identical(isTRUE(item$hasR), has_r), logical(1))
  }

  items[keep]
}

get_ngplot_template <- function(template_id, base_dir = NULL) {
  items <- list_ngplot_templates(base_dir)
  matched <- Filter(function(item) identical(item$id, template_id), items)
  if (length(matched) == 0) {
    stop(sprintf("Template not found: %s", template_id), call. = FALSE)
  }
  matched[[1]]
}

read_ngplot_bundle <- function(template_id, base_dir = NULL) {
  base <- ngplot_base_dir(base_dir)
  item <- get_ngplot_template(template_id, base)
  bundle <- item
  bundle$meta <- fromJSON(file.path(base, item$metaUrl), simplifyVector = FALSE)
  bundle$svg <- readChar(file.path(base, item$svgUrl), file.info(file.path(base, item$svgUrl))$size, useBytes = TRUE)
  bundle$data <- setNames(lapply(item$dataFiles, function(file_info) {
    readChar(file.path(base, file_info$url), file.info(file.path(base, file_info$url))$size, useBytes = TRUE)
  }), vapply(item$dataFiles, function(file_info) file_info$name, character(1)))
  if (!is.null(item$pythonUrl) && nzchar(item$pythonUrl)) {
    bundle$python_code <- readChar(file.path(base, item$pythonUrl), file.info(file.path(base, item$pythonUrl))$size, useBytes = TRUE)
  }
  if (!is.null(item$rUrl) && nzchar(item$rUrl)) {
    bundle$r_code <- readChar(file.path(base, item$rUrl), file.info(file.path(base, item$rUrl))$size, useBytes = TRUE)
  }
  bundle
}

File: bioinforw-data-plotting/assets/bioinforw_ngplot_dump/catalog_loader.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def _base_dir(base_dir: str | Path | None = None) -> Path:
    return Path(base_dir) if base_dir is not None else Path(__file__).resolve().parent


def _catalog_path(base_dir: str | Path | None = None) -> Path:
    return _base_dir(base_dir) / "catalog.json"


def load_catalog(base_dir: str | Path | None = None) -> dict[str, Any]:
    return json.loads(_catalog_path(base_dir).read_text(encoding="utf-8"))


def list_templates(base_dir: str | Path | None = None) -> list[dict[str, Any]]:
    return load_catalog(base_dir)["templates"]


def find_templates(
    keyword: str | None = None,
    category: str | None = None,
    has_python: bool | None = None,
    has_r: bool | None = None,
    base_dir: str | Path | None = None,
) -> list[dict[str, Any]]:
    items = list_templates(base_dir)
    if keyword:
        needle = keyword.lower()
        items = [
            item
            for item in items
            if needle in item["name"].lower()
            or needle in item["id"].lower()
            or needle in item["category"].lower()
            or needle in item["categoryTrailText"].lower()
        ]
    if category:
        items = [item for item in items if item["category"] == category]
    if has_python is not None:
        items = [item for item in items if bool(item["hasPython"]) is has_python]
    if has_r is not None:
        items = [item for item in items if bool(item["hasR"]) is has_r]
    return items


def get_template(template_id: str, base_dir: str | Path | None = None) -> dict[str, Any]:
    for item in list_templates(base_dir):
        if item["id"] == template_id:
            return item
    raise KeyError(f"Template not found: {template_id}")


def load_bundle(template_id: str, base_dir: str | Path | None = None) -> dict[str, Any]:
    root = _base_dir(base_dir)
    item = get_template(template_id, root)
    bundle: dict[str, Any] = dict(item)
    bundle["meta"] = json.loads((root / item["metaUrl"]).read_text(encoding="utf-8"))
    bundle["svg"] = (root / item["svgUrl"]).read_text(encoding="utf-8")
    bundle["data"] = {
        file_info["name"]: (root / file_info["url"]).read_text(encoding="utf-8")
        for file_info in item["dataFiles"]
    }
    if item.get("pythonUrl"):
        bundle["python_code"] = (root / item["pythonUrl"]).read_text(encoding="utf-8")
    if item.get("rUrl"):
        bundle["r_code"] = (root / item["rUrl"]).read_text(encoding="utf-8")
    return bundle


if __name__ == "__main__":
    catalog = load_catalog()
    print(f"Loaded {catalog['totalTemplates']} templates from {_catalog_path()}")
